src/e84_geoai_common/llm/batch.py

```diff
@@ -0,0 +1,324 @@
+        response = self.s3_client.get_object(Bucket=output_bucket, Key=key)
+        response_body = response["Body"].read().decode("utf-8")
+        output_messages = [
+            self.batch_record_output_model.model_validate_json(line)
+            for line in response_body.splitlines()
+        ]
+
+        return output_messages
+
+    def wait_for_job_to_finish(self: Self, job_arn: str) -> None:
+        """Returns once the job has either finished or failed."""
+        while True:
+            response = self.bedrock_client.get_model_invocation_job(jobIdentifier=job_arn)
+            status = response.get("status")
+
+            if status == "Completed":
+                break
+
+            if status in ["Failed", "Stopping", "Stopped"]:
+                message = response.get("message")
+                error_details = (
+                    f"Job {job_arn} finished with status: {status}. Message: '{message}'."
+                )
+                raise RuntimeError(error_details)
+
+            sleep(30)
+
+    def get_job_arn(self: Self, job_name: str) -> str:
+        """Returns job arn given the job name."""
+        paginator = self.bedrock_client.get_paginator("list_model_invocation_jobs")
+
+        page_iterator = paginator.paginate(
+            nameContains=job_name, sortOrder="Descending", sortBy="CreationTime"
+        )
+
+        job_arn_found = None
+        for page in page_iterator:
+            for job_summary in page.get("invocationJobSummaries", []):
+                if job_summary.get("jobName") == job_name:
+                    job_arn_found = job_summary.get("jobArn")
+                    break
+            if job_arn_found:
+                break
+
+        if not job_arn_found:
+            msg = f"Bedrock Batch Job with name '{job_name}' not found)"
+            raise ValueError(msg)
+
+        return job_arn_found
+
+    def create_job(  # noqa: PLR0913
+        self: Self,
+        job_name: str,
+        role_arn: str,
+        input_s3_file_url: str,
+        output_s3_directory_url: str,
+        conversations: list[list[LLMMessage]] | None = None,
+        inference_cfg: LLMInferenceConfig | None = None,
+        *,
+        create_buckets_if_missing: bool,
+    ) -> str:
+        """Creates and invokes batch job."""
+        # check to make sure that these parse and error check correctly
+        input_bucket, input_bucket_key, output_bucket = self._parse_s3_urls(
+            input_s3_file_url, output_s3_directory_url
+        )
+
+        self._ensure_bucket_exists(
+            input_bucket,
+            create_buckets_if_missing=create_buckets_if_missing,
+        )
+        self._ensure_bucket_exists(
+            output_bucket,
+            create_buckets_if_missing=create_buckets_if_missing,
+        )
+
+        # Upload files if conversations exist
+        if conversations:
+            if inference_cfg is None:
+                inference_cfg = LLMInferenceConfig()
```

Comment:
I like that inference_cfg is optional. I find whenever I use the library I almost never want to customize it and have to create one. We should probably do that on our other APIs in some future PR. CC @AdeelH